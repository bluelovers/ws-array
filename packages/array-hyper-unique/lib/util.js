"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultChecker = exports.defaultFilter = exports.equals = void 0;
const deep_eql_1 = __importDefault(require("deep-eql"));
function equals(a1, a2) {
    return deep_eql_1.default(a1, a2);
}
exports.equals = equals;
function defaultFilter(options = {}) {
    const checker = options.checker || defaultChecker;
    const filter = options.filter || null;
    const cb = (val, index, arr) => {
        let i = arr.findIndex(a => checker(a, val, arr, arr));
        return i == index && (!filter || filter(val));
    };
    return cb;
}
exports.defaultFilter = defaultFilter;
function defaultChecker(element, value, arr_new, arr_old) {
    return deep_eql_1.default(element, value);
}
exports.defaultChecker = defaultChecker;
//# sourceMappingURL=util.js.map