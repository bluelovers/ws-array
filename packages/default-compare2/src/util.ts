
export function _typeOf(target: unknown)
{
	if (target === null)
	{
		return 'null' as const
	}

	return typeof target
}
