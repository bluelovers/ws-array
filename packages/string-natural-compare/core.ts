/**
 * Created by user on 2020/6/4.
 */

import _naturalCompare from 'string-natural-compare';

export interface IOptionsStringNaturalCompare
{
	/**
	 * Set to true to compare strings case-insensitively. Default: false.
	 */
	caseInsensitive?: boolean,
	/**
	 * A string of characters that define a custom character ordering. Default: undefined.
	 */
	alphabet?: string,
}

export interface IOptionsNaturalCompare extends IOptionsStringNaturalCompare
{
	desc?: boolean,
}

/**
 * Compare alphanumeric strings the same way a human would,
 * using a natural order algorithm
 * (originally known as the alphanum algorithm)
 * where numeric characters are sorted
 * based on their numeric values rather than their ASCII values.
 */
export function naturalCompare(a, b, opts?: IOptionsNaturalCompare)
{
	let i: number;

	if (typeof a === 'number' && typeof b === 'number')
	{
		i = a - b
	}
	else
	{
		if (typeof a === 'number')
		{
			a = String(a);
		}
		else if (typeof b === 'number')
		{
			b = String(b);
		}

		if (a === b)
		{
			return 0
		}

		i = _naturalCompare(a, b, opts)
	}

	if (i !== 0 && opts?.desc)
	{
		i = 0 - i;
	}

	return i
}

/**
 * create compare with preset options
 */
export function createNew(opts?: IOptionsNaturalCompare)
{
	return (a, b) => naturalCompare(a, b, opts)
}

/**
 * compare strings case-insensitively
 */
export const compareCaseInsensitive = createNew({
	caseInsensitive: true,
})

naturalCompare.createNew = createNew;
naturalCompare.compareCaseInsensitive = compareCaseInsensitive;
naturalCompare.default = naturalCompare;

export default naturalCompare
