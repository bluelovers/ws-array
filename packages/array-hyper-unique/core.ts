import { defaultFilter, equals, defaultChecker } from './lib/util';
import { IOptions } from './lib/types';

export * from './lib/types';
export * from './lib/util';

export function array_unique<T>(arr: T, options: IOptions<T> = {}): T
{
	if (!Array.isArray(arr))
	{
		throw new TypeError(`Expected an Array but got ${typeof arr}.`)
	}

	const cb = defaultFilter(options);

	if (options.overwrite)
	{
		let index = arr.length;

		while (index--)
		{
			let val = arr[index];

			if (!cb(val, index, arr))
			{
				arr.splice(index, 1);
			}
		}

		return arr;
	}

	// @ts-ignore
	return arr.filter(cb);
}

export function array_unique_overwrite<T>(arr: T, options: IOptions<T> = {}): T
{
	return array_unique(arr, {
		...options,
		overwrite: true,
	});
}

export function lazy_unique<T extends any[]>(arr: T): T
export function lazy_unique<T, T1, T2>(a1: T1, a2: T2, ...arr: T[]): Array<T | T1 | T2>
export function lazy_unique<T>(...arr: Array<T | T[]>): T | (T | T[])[]
// @ts-ignore
export function lazy_unique<T>(...arr: Array<T | T[]>)
{
	if (arr.length > 1)
	{
		return array_unique(arr);
	}

	return array_unique(arr[0]);
}

export function lazy_unique_overwrite<T>(...arr: Array<T | T[]>)
{
	if (arr.length > 1)
	{
		return array_unique_overwrite(arr);
	}

	return array_unique_overwrite(arr[0]);
}

lazy_unique.array_unique = array_unique;
lazy_unique.array_unique_overwrite = array_unique_overwrite;

lazy_unique.lazy_unique_overwrite = lazy_unique_overwrite;

lazy_unique.equals = equals;
lazy_unique.defaultFilter = defaultFilter;
lazy_unique.defaultChecker = defaultChecker;

lazy_unique.lazy_unique = lazy_unique;
lazy_unique.default = lazy_unique;

Object.defineProperty(lazy_unique, "__esModule", { value: true });

export default lazy_unique
