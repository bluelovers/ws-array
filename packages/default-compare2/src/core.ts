import { _typeOf } from './util';

export const enum EnumSortCompareResult
{
	KEEP = 0,
	RIGHT = 1,
	LEFT = -1,
}

export function numberCompare(a: unknown, b: unknown)
{
	if (a < b)
	{
		return EnumSortCompareResult.LEFT
	}
	else if (a > b)
	{
		return EnumSortCompareResult.RIGHT
	}

	return EnumSortCompareResult.KEEP
}

/**
 * will return undefined when a and b not null/undefined
 */
export function nullCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'null')
	{
		return typeB === 'null' ? EnumSortCompareResult.KEEP : (typeB === 'undefined'
			? EnumSortCompareResult.LEFT
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeA === 'undefined')
	{
		return typeB === 'null' ? EnumSortCompareResult.RIGHT : (typeB === 'undefined'
			? EnumSortCompareResult.KEEP
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeB === 'null' || typeB === 'undefined')
	{
		return EnumSortCompareResult.LEFT;
	}
}

export function symbolCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'symbol')
	{
		return typeB === 'symbol' ? numberCompare(String(a), String(b)) : EnumSortCompareResult.LEFT
	}
	else if (typeB === 'symbol')
	{
		return EnumSortCompareResult.RIGHT
	}

	return numberCompare(a, b)
}

export function dateCompare(a: Date, b: Date)
{
	return a.getTime() - b.getTime()
}

export function stringCompareLocale(a: string, b: string)
{
	return a.localeCompare(b)
}
