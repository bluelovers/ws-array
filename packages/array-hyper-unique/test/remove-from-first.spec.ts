import getTestcase from './_data';
import * as core from '../core';
import { inspect } from 'util';

describe(`removeFromFirst`, () =>
{

	let fn1 = function() {};
	let fn2 = function() {};
	let fn3 = function() {};


	([
		{
			label: `simple`,
			data: [
				2,
				4,
				3,
				1,
				3,
				2,
			],
		},
		{
			label: `mixin test`,
			data: [
				2,
				4,
				1,
				{ a: {} },
				'undefined',
				'null',
				'false',
				'',
				NaN, NaN, NaN, Infinity,
				fn1, fn2, fn3, fn2, fn3, fn2,
				3,
				1,
				3,
				['a', 'b', 'c'],
				['a', 'c', 'b'],
				{ a: { b: 2 } },

				{ a: { b: 3, c: undefined } },
				/1/,
				{},
				/1/,
				/1/g,
				0,
				true,
				undefined,
				null,
				false,
				['a', 'b', 'c'],
				['a', 'b', 'c'],
				['a', 'c', 'b'],
				{ a: { b: 2 } },

				/1/,
				{},
				/1/,
				/1/g,
				/1/i,
				{ a: { b: 2 } },
				{ a: { b: 3 } },
				{ a: { b: undefined } },
				{ a: {} },
				{ a: { b: 3, c: undefined } },
				{},
				1,
				0,
				true,
				undefined,
				null,
				false,
				['a', 'b', 'c'],

				'true',
				{ a: { b: 2 } },
				{ a: { b: 3 } },
				{ a: { b: undefined } },

				/1/i,
				{},
				/1/,
				/1/g,
				{},
				'1',
				'0',
				2,
			],
		},
	])
		.forEach(function ({
			data,
			label,
		}, index)
		{
			_testIt(data, index, `array_unique`, label)
			_testIt(data, index, `array_unique_overwrite`, label)

		})
	;

	function _testIt(data, index, fnName: keyof typeof core, label?: string)
	{
		label = label ?? `${index} array(${data.length})`;
		label = `${fnName} ${label}`

		it(label, (done) =>
		{
			const fn = core[fnName]

			const len = data.length;

			// @ts-ignore
			let a1 = fn(data.slice())
			// @ts-ignore
			let a2 = fn(data.slice(), {
				removeFromFirst: true,
			})

			expect(a1.length).toBeLessThan(len);

			expect(a1).toHaveLength(a2.length);

			expect(a1).not.toStrictEqual(data);
			expect(a2).not.toStrictEqual(data);

			expect(a1).not.toStrictEqual(a2);

			expect({
				a1,
				a2,
			}).toMatchSnapshot();

			//a1.sort();
			//a2.sort();

			//expect(a1).toStrictEqual(a2);

			//expect(actual).toMatchSnapshot();

			done();
		});
	}

})
