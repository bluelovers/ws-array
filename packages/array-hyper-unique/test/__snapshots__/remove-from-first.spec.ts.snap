// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeFromFirst array_unique mixin test 1`] = `
Object {
  "a1": Array [
    2,
    4,
    1,
    Object {
      "a": Object {},
    },
    "undefined",
    "null",
    "false",
    "",
    NaN,
    Infinity,
    [Function],
    [Function],
    [Function],
    3,
    Array [
      "a",
      "b",
      "c",
    ],
    Array [
      "a",
      "c",
      "b",
    ],
    Object {
      "a": Object {
        "b": 2,
      },
    },
    Object {
      "a": Object {
        "b": 3,
        "c": undefined,
      },
    },
    /1/,
    Object {},
    /1/g,
    0,
    true,
    undefined,
    null,
    false,
    /1/i,
    Object {
      "a": Object {
        "b": 3,
      },
    },
    Object {
      "a": Object {
        "b": undefined,
      },
    },
    "true",
    "1",
    "0",
  ],
  "a2": Array [
    4,
    "undefined",
    "null",
    "false",
    "",
    NaN,
    Infinity,
    [Function],
    [Function],
    [Function],
    3,
    Array [
      "a",
      "c",
      "b",
    ],
    Object {
      "a": Object {},
    },
    Object {
      "a": Object {
        "b": 3,
        "c": undefined,
      },
    },
    1,
    0,
    true,
    undefined,
    null,
    false,
    Array [
      "a",
      "b",
      "c",
    ],
    "true",
    Object {
      "a": Object {
        "b": 2,
      },
    },
    Object {
      "a": Object {
        "b": 3,
      },
    },
    Object {
      "a": Object {
        "b": undefined,
      },
    },
    /1/i,
    /1/,
    /1/g,
    Object {},
    "1",
    "0",
    2,
  ],
}
`;

exports[`removeFromFirst array_unique simple 1`] = `
Object {
  "a1": Array [
    2,
    4,
    3,
    1,
  ],
  "a2": Array [
    4,
    1,
    3,
    2,
  ],
}
`;

exports[`removeFromFirst array_unique_overwrite mixin test 1`] = `
Object {
  "a1": Array [
    2,
    4,
    1,
    Object {
      "a": Object {},
    },
    "undefined",
    "null",
    "false",
    "",
    NaN,
    Infinity,
    [Function],
    [Function],
    [Function],
    3,
    Array [
      "a",
      "b",
      "c",
    ],
    Array [
      "a",
      "c",
      "b",
    ],
    Object {
      "a": Object {
        "b": 2,
      },
    },
    Object {
      "a": Object {
        "b": 3,
        "c": undefined,
      },
    },
    /1/,
    Object {},
    /1/g,
    0,
    true,
    undefined,
    null,
    false,
    /1/i,
    Object {
      "a": Object {
        "b": 3,
      },
    },
    Object {
      "a": Object {
        "b": undefined,
      },
    },
    "true",
    "1",
    "0",
  ],
  "a2": Array [
    4,
    "undefined",
    "null",
    "false",
    "",
    NaN,
    Infinity,
    [Function],
    [Function],
    [Function],
    3,
    Array [
      "a",
      "c",
      "b",
    ],
    Object {
      "a": Object {},
    },
    Object {
      "a": Object {
        "b": 3,
        "c": undefined,
      },
    },
    1,
    0,
    true,
    undefined,
    null,
    false,
    Array [
      "a",
      "b",
      "c",
    ],
    "true",
    Object {
      "a": Object {
        "b": 2,
      },
    },
    Object {
      "a": Object {
        "b": 3,
      },
    },
    Object {
      "a": Object {
        "b": undefined,
      },
    },
    /1/i,
    /1/,
    /1/g,
    Object {},
    "1",
    "0",
    2,
  ],
}
`;

exports[`removeFromFirst array_unique_overwrite simple 1`] = `
Object {
  "a1": Array [
    2,
    4,
    3,
    1,
  ],
  "a2": Array [
    4,
    1,
    3,
    2,
  ],
}
`;
